proxy/src/proxy.rs

```diff
@@ -59,6 +61,23 @@ static NUM_BYTES_PROXIED_COUNTER: Lazy<IntCounterVec> = Lazy::new(|| {
     .unwrap()
 });
 
+static EXIT_ON_ALL_CONNECTIONS_CLOSED: AtomicBool = AtomicBool::new(false);
+
+pub fn set_exit_on_connections_closed() {
+    EXIT_ON_ALL_CONNECTIONS_CLOSED.store(true, Ordering::Release);
+}
+
+pub fn exit_if_needed() {
+    if EXIT_ON_ALL_CONNECTIONS_CLOSED.load(Ordering::Acquire) {
+        let num_accepted = NUM_CONNECTIONS_ACCEPTED_COUNTER.get();
+        let num_closed = NUM_CONNECTIONS_CLOSED_COUNTER.get();
+        if num_accepted == num_closed {
+            info!("All connections closed, terminating");
+            process::exit(0);
```

Comment:
I decided cleanest way would just be to have a static CancellationToken in `proxy` which gets cancelled on `exit_if_needed`. It would be cleaner to have token be created in `proxy::task_main` and passed around, except having the `handle_signals` function also have a copy of the token is annoying.